File: packages/core/src/fileIcons.ts
// Ícone por nome de arquivo/pasta, no tema do VS Code (material-icon-theme). Só o subconjunto de
// fileIcons.lista.json existe no bundle; o resto cai no genérico — que é o `document`/`folder`
// do próprio tema, não os dois SVGs de traço de antes.
import { ICONES } from './fileIcons.generated';

const PASTA: Record<string, string> = {
  src: 'src', app: 'src',
  pages: 'views', views: 'views', screens: 'views', routes: 'views',
  components: 'components', widgets: 'components',
  lib: 'lib', libs: 'lib', utils: 'lib', helpers: 'lib',
  test: 'test', tests: 'test', __tests__: 'test', spec: 'test', e2e: 'test',
  docs: 'docs', doc: 'docs',
  node_modules: 'node',
  '.github': 'github', '.vscode': 'vscode',
  scripts: 'scripts', bin: 'scripts', hooks: 'scripts',
  public: 'public', static: 'public', assets: 'public',
  dist: 'dist', build: 'dist', out: 'dist',
  api: 'api', backend: 'api', server: 'api',
  images: 'images', img: 'images', icons: 'images',
  config: 'config', '.config': 'config', settings: 'config',
  '.git': 'git',
};

const NOME: Record<string, string> = {
  'package.json': 'nodejs', 'package-lock.json': 'lock', 'pnpm-lock.yaml': 'lock', 'yarn.lock': 'lock', 'uv.lock': 'lock',
  'claude.md': 'markdown', 'dockerfile': 'docker', '.gitignore': 'git', '.gitattributes': 'git',
  '.editorconfig': 'editorconfig', '.prettierrc': 'prettier', 'biome.json': 'biome',
  'vitest.config.ts': 'vitest', 'jest.config.js': 'jest', 'pyproject.toml': 'toml',
};

const EXT: Record<string, string> = {
  svelte: 'svelte', ts: 'typescript', mts: 'typescript', cts: 'typescript', tsx: 'react', jsx: 'react',
  js: 'javascript', mjs: 'javascript', cjs: 'javascript', py: 'python', cs: 'csharp', dart: 'dart',
  pas: 'pascal', dpr: 'pascal', dfm: 'pascal', md: 'markdown', mdx: 'markdown', json: 'json', jsonc: 'json',
  css: 'css', scss: 'css', less: 'css', html: 'html', htm: 'html', xml: 'xml', svg: 'image',
  sql: 'database', db: 'database', sqlite: 'database', sh: 'console', bash: 'console', fish: 'console',
  zsh: 'console', ps1: 'powershell', yaml: 'yaml', yml: 'yaml', toml: 'toml', lock: 'lock',
  png: 'image', jpg: 'image', jpeg: 'image', gif: 'image', webp: 'image', avif: 'image', bmp: 'image', ico: 'image',
  pdf: 'pdf', txt: 'document', log: 'document', zip: 'zip', tgz: 'zip', gz: 'zip', tar: 'zip', '7z': 'zip',
  ttf: 'font', otf: 'font', woff: 'font', woff2: 'font', mp4: 'video', mov: 'video', webm: 'video', mkv: 'video',
  mp3: 'audio', wav: 'audio', m4a: 'audio', ogg: 'audio', env: 'tune',
};

// ponytail: prefixo testado por startsWith — cobre .env.local, README.pt.md, tsconfig.app.json,
// vite.config.ts, docker-compose.dev.yaml sem regex.
const PREFIXO: [string, string][] = [
  ['.env', 'tune'], ['readme', 'readme'], ['tsconfig', 'tsconfig'], ['vite.config', 'vite'],
  ['docker-compose', 'docker'], ['.eslintrc', 'eslint'], ['eslint.config', 'eslint'],
  ['.prettierrc', 'prettier'], ['vitest.config', 'vitest'], ['settings.', 'settings'],
];

export function nomeIcone(nome: string, isDir: boolean, aberta = false): string {
  const n = nome.toLowerCase();
  if (isDir) {
    const p = PASTA[n];
    return p ? `folder-${p}${aberta ? '-open' : ''}` : (aberta ? 'folder-open' : 'folder');
  }
  const direto = NOME[n] ?? PREFIXO.find(([pre]) => n.startsWith(pre))?.[1];
  if (direto) return direto;
  const i = n.lastIndexOf('.');
  const ext = i >= 0 ? n.slice(i + 1) : '';
  return EXT[ext] ?? 'document';
}

export function svgIcone(nome: string, isDir: boolean, aberta = false): string {
  return ICONES[nomeIcone(nome, isDir, aberta)] ?? ICONES[isDir ? 'folder' : 'document'];
}

File: desktop-native/src/fileicons.rs
use gpui_kit::{img, Img, SharedString, Styled};

// Mesma prioridade do core/fileIcons.ts: nome, prefixo, extensão, genérico.
fn icon_name(name: &str, is_dir: bool, open: bool) -> String {
    let name = name.to_lowercase();
    if is_dir {
        let folder = match name.as_str() {
            "src" | "app" => "src",
            "pages" | "views" | "screens" | "routes" => "views",
            "components" | "widgets" => "components",
            "lib" | "libs" | "utils" | "helpers" => "lib",
            "test" | "tests" | "__tests__" | "spec" | "e2e" => "test",
            "docs" | "doc" => "docs",
            "node_modules" => "node",
            ".github" => "github",
            ".vscode" => "vscode",
            "scripts" | "bin" | "hooks" => "scripts",
            "public" | "static" | "assets" => "public",
            "dist" | "build" | "out" => "dist",
            "api" | "backend" | "server" => "api",
            "images" | "img" | "icons" => "images",
            "config" | ".config" | "settings" => "config",
            ".git" => "git",
            _ => "",
        };
        return format!("folder{}{}{}", if folder.is_empty() { "" } else { "-" }, folder, if open { "-open" } else { "" });
    }
    let exact = match name.as_str() {
        "package.json" => "nodejs",
        "package-lock.json" | "pnpm-lock.yaml" | "yarn.lock" | "uv.lock" => "lock",
        "claude.md" => "markdown",
        "dockerfile" => "docker",
        ".gitignore" | ".gitattributes" => "git",
        ".editorconfig" => "editorconfig",
        ".prettierrc" => "prettier",
        "biome.json" => "biome",
        "vitest.config.ts" => "vitest",
        "jest.config.js" => "jest",
        "pyproject.toml" => "toml",
        _ => "",
    };
    if !exact.is_empty() { return exact.to_owned(); }
    for (prefix, icon) in [
        (".env", "tune"),
        ("readme", "readme"),
        ("tsconfig", "tsconfig"),
        ("vite.config", "vite"),
        ("docker-compose", "docker"),
        (".eslintrc", "eslint"),
        ("eslint.config", "eslint"),
        (".prettierrc", "prettier"),
        ("vitest.config", "vitest"),
        ("settings.", "settings"),
    ] {
        if name.starts_with(prefix) { return icon.to_owned(); }
    }
    match name.rsplit_once('.').map_or("", |(_, ext)| ext) {
        "svelte" => "svelte",
        "ts" | "mts" | "cts" => "typescript",
        "tsx" | "jsx" => "react",
        "js" | "mjs" | "cjs" => "javascript",
        "py" => "python",
        "cs" => "csharp",
        "dart" => "dart",
        "pas" | "dpr" | "dfm" => "pascal",
        "md" | "mdx" => "markdown",
        "json" | "jsonc" => "json",
        "css" | "scss" | "less" => "css",
        "html" | "htm" => "html",
        "xml" => "xml",
        "svg" | "png" | "jpg" | "jpeg" | "gif" | "webp" | "avif" | "bmp" | "ico" => "image",
        "sql" | "db" | "sqlite" => "database",
        "sh" | "bash" | "fish" | "zsh" => "console",
        "ps1" => "powershell",
        "yaml" | "yml" => "yaml",
        "toml" => "toml",
        "lock" => "lock",
        "pdf" => "pdf",
        "txt" | "log" => "document",
        "zip" | "tgz" | "gz" | "tar" | "7z" => "zip",
        "ttf" | "otf" | "woff" | "woff2" => "font",
        "mp4" | "mov" | "webm" | "mkv" => "video",
        "mp3" | "wav" | "m4a" | "ogg" => "audio",
        "env" => "tune",
        _ => "document",
    }.to_owned()
}

pub fn citation_icon(name: &str) -> Img {
    // img preserva as cores dos SVGs; svg/Icon os converte em máscara monocromática.
    img(SharedString::from(format!("file-icons/{}.svg", icon_name(name, false, false))))
        .size_3p5().flex_shrink_0()
}

macro_rules! file_assets {
    ($($name:literal),* $(,)?) => {
        const ASSETS: &[(&str, &[u8])] = &[$(
            (concat!("file-icons/", $name, ".svg"), include_bytes!(concat!("../assets/file-icons/", $name, ".svg"))),
        )*];
    };
}

file_assets!(
    "folder", "folder-open", "folder-src", "folder-src-open", "folder-views",
    "folder-views-open", "folder-components", "folder-components-open", "folder-lib", "folder-lib-open",
    "folder-test", "folder-test-open", "folder-docs", "folder-docs-open", "folder-node",
    "folder-node-open", "folder-github", "folder-github-open", "folder-vscode", "folder-vscode-open",
    "folder-scripts", "folder-scripts-open", "folder-public", "folder-public-open", "folder-dist",
    "folder-dist-open", "folder-api", "folder-api-open", "folder-images", "folder-images-open",
    "folder-config", "folder-config-open", "folder-git", "folder-git-open", "svelte",
    "typescript", "javascript", "react", "python", "csharp",
    "dart", "pascal", "markdown", "json", "css",
    "html", "xml", "database", "console", "powershell",
    "yaml", "toml", "lock", "image", "pdf",
    "document", "zip", "font", "video", "audio",
    "tune", "nodejs", "readme", "tsconfig", "vite",
    "biome", "git", "docker", "settings", "eslint",
    "prettier", "editorconfig", "vitest", "jest",
);

pub fn load(path: &str) -> Option<&'static [u8]> {
    ASSETS.iter().find(|(name, _)| *name == path).map(|(_, bytes)| *bytes)
}

#[cfg(test)]
mod tests {
    use super::*;
    use gpui_kit::AssetSource;

    #[test]
    fn web_priority_fallbacks_and_embedded_assets() {
        for (name, expected) in [
            ("main.rs", "document"), ("main.TS", "typescript"), ("note.md", "markdown"),
            ("photo.png", "image"), ("unknown.xyz", "document"), ("Makefile", "document"),
            ("package.json", "nodejs"), ("package-lock.json", "lock"), ("Dockerfile", "docker"),
            ("CLAUDE.md", "markdown"), ("README.pt.md", "readme"), (".env.local", "tune"),
            ("tsconfig.app.json", "tsconfig"), ("docker-compose.dev.yaml", "docker"),
            ("vitest.config.ts", "vitest"), ("settings.json", "settings"), ("src", "document"),
            ("view.tsx", "react"), ("module.mts", "typescript"), ("form.dfm", "pascal"),
        ] {
            assert_eq!(icon_name(name, false, false), expected, "{name}");
        }
        assert_eq!(icon_name("SRC", true, false), "folder-src");
        assert_eq!(icon_name("src", true, true), "folder-src-open");
        assert_eq!(icon_name("other", true, false), "folder");
        assert_eq!(icon_name("other", true, true), "folder-open");
        for (path, bytes) in ASSETS {
            assert!(bytes.starts_with(b"<svg"), "{path}");
            assert_eq!(crate::AppAssets.load(path).unwrap().unwrap().as_ref(), *bytes, "{path}");
        }
        assert!(load("file-icons/missing.svg").is_none());
    }
}
